resolwe/flow/views/entity.py

```diff
@@ -49,14 +51,60 @@ def _get_entities(self, user, ids):
 
     @action(detail=False, methods=["post"])
     def move_to_collection(self, request, *args, **kwargs):
-        """Move samples from source to destination collection."""
+        """Move samples from source to destination collection.
+
+        The request accepts argument 'handle_missing_annotations'.The possible
+        values for it are 'ADD' and 'REMOVE'. When 'REMOVE' is set without
+        arguments the 'confirm_annotations_delete' argument must be set to
+        True.
+
+        :raises ValidationError: when handle_missing_annotations has incorrect
+            value.
+        :raises ValidationError: when handle_missing_annotations is set to
+            'REMOVE' and 'confirm_annotations_delete' is set.
+        """
+
         ids = self.get_ids(request.data)
         dst_collection_id = self.get_id(request.data, "destination_collection")
         dst_collection = get_collection_for_user(dst_collection_id, request.user)
-
         entity_qs = self._get_entities(request.user, ids)
-        entity_qs.move_to_collection(dst_collection)
 
+        missing_annotations = {
+            entity: AnnotationField.objects.filter(collections=entity.collection)
+            .exclude(collections=dst_collection)
+            .distinct()
+            for entity in entity_qs
+        }
+
+        if missing_annotations:
+            # How to handle annotation fields present in the current collection
+            # but missing in new one. We have two options:
+            # - add (default): add the missing annotation fields to the new
+            #   collection and keep the values.
+            # - delete: remove the missing annotation values from the samples.
+            #   This option requires explicit confirmation by 'confirm_remove'
+            #   argument or exception with explanation is raised.
+            try:
+                handle_missing_annotations = HandleMissingAnnotations(
+                    request.data.get(
+                        "handle_missing_annotations", HandleMissingAnnotations.ADD.name
+                    )
+                )
+            except ValueError:
+                possible_values = HandleMissingAnnotations._member_names_
+                raise ValidationError(
+                    "Attribute 'handle_missing_annotations' must be one of "
+                    f"{possible_values}."
+                )
+
+            if handle_missing_annotations == HandleMissingAnnotations.REMOVE:
+                if request.data.get("confirm_annotations_remove", False) is not True:
+                    raise ValidationError(
+                        "All annotations not present in the target collection "
+                        "will be removed. Please confirm."
```

Comment:
How to confirm?

Reply:
I am not sure how to fix the message. Adding implementation detail (set `confirm_annotations_remove` to True)  seems a little weird.

When creating the message my assumption was that the message is shown to the user by the frontend along with the confirmation button or something similar.

Reply:
This will be used by different clients, so you cannot do such assumptions. I think you can write a bit more technical message about how the client should handle the confirmation, not how the user should do it.

Reply:
Done.